### ISSUE TITLE
Ignore multi-line calls in duplicate detection (R0801)

### ISSUE THREAD

**Issue by Bengt** [PyCQA/pylint#7920]
### Current problem

When I run PyLint on my code base, which contains many multi-line function calls like this:

```python
return_value_0,
return_value_1,
return_value_2,
return_value_3 = \
    my_function(
        parameter_0,
        parameter_1,
        parameter_2,
        parameter_3,
)

PyLint complains over duplicate code:

    R0801: Similar lines in 2 files

Note that while this call spans multiple lines, it is only a single statement. Maybe single statements should never be considered duplicates, as they might just be the canonical form?

### Desired solution

I would like PyLint to detect these multi-line duplicates as okay, since they are the canonical way of calling these functions in my style of formatting the code.

### Additional context

@Pierre-Sassoulas suggested introducing an option like `--ignore-multiline-calls` (like other similarities checker options https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#similarities-checker) here: https://github.com/PyCQA/pylint/issues/214#issuecomment-1345634225

**Comment by moomingoppe** [comment]
I would like this as well. We often find this issue since we re-use the same method many times

**Comment by Pierre-Sassoulas** [comment]
I'm -0.1 on this one because it's actual duplication and it can be fixed by using data structure. or classes. i.e. ``legs, torso, arms = muscle_up(leg, torso, arms)`` => ``dog.muscle_up()`` or ``superdog = muscle_up(dog)``. Which was not the case for the signatures option where you had to implement the API.

**Comment by ruro** [comment]
> it's actual duplication and it can be fixed by using data structure

I disagree. It can be **worked around**, sure, but it can't really be "fixed". "Fix" implies that there is some problem with the code. In the `dog` example that you gave, all the arguments and return values are strongly coupled together, so it makes sense to refactor them into a separate class. However, this won't always be the case.

This warning doesn't only happen in the "nice case" where refactoring the arguments/return values makes sense and is easy to do. In fact, I think that the whole function call is actually a red herring in this case. From my limited testing, it seems that this problem is actually caused by the "long" destructuring assignment ALONE. And destructuring the return value of a function just happens to be a common case where this issue presents itself.

For example, this MRE also triggers the same warning:
```sh
mkdir -p example
touch example/__init__.py

cat >example/common.py <<'EOF'
SOME_DATA = 1, 2, 3, 4
EOF

cat >example/bar1.py <<'EOF'
from .common import SOME_DATA


class Bar1:
    def different(self):
        return 0

    def bee(self):
        (
            a_lot_of_local_variables,
            with_long_names,
            that_dont_fit,
            on_one_line,
        ) = SOME_DATA

        print("this is completely different code")
        that_dont_fit -= with_long_names
        print(a_lot_of_local_variables)
        print(that_dont_fit)
        return on_one_line ** with_long_names
EOF

cat >example/bar2.py <<'EOF'
from .common import SOME_DATA


class Bar2:
    def woop(self):
        print("some different code here")

        (
            a_lot_of_local_variables,
            with_long_names,
            that_dont_fit,
            on_one_line,
        ) = SOME_DATA

        some_calculation = a_lot_of_local_variables + with_long_names
        print(some_calculation, on_one_line / that_dont_fit)

    def unrelated(self):
        return "clearly different"
EOF

pylint --rcfile= --disable=missing-docstring example
```

```
************* Module example.bar2
example/bar2.py:1:0: R0801: Similar lines in 2 files
==example.bar1:[8:15]
==example.bar2:[7:14]
        (
            a_lot_of_local_variables,
            with_long_names,
            that_dont_fit,
            on_one_line,
        ) = SOME_DATA
 (duplicate-code)

------------------------------------------------------------------
Your code has been rated at 9.52/10 (previous run: 9.52/10, +0.00)
```

I have a tuple with some initial values and I want to initialize a bunch of local variables using that data. How is this duplication?

And how would you "fix" this? Extracting those values into a class doesn't help, because you will still eventually need to perform the assignment to those local variables and IMHO having to repeat something like
```py
a_lot_of_local_variables = SOME_OBJECT.a_lot_of_local_variables
with_long_names = SOME_OBJECT.with_long_names
that_dont_fit = SOME_OBJECT.that_dont_fit
on_one_line = SOME_OBJECT.on_one_line
```
is worse than the destructuring assignment, not better.

**Comment by kanrrra** [comment]
> it's actual duplication and it can be fixed by using data structure

I agree with @ruro that this is sometimes an improvement but not always. You dont always have control over the function you are calling. For example with something common like the requests library, you may have something like:

```
url = "https://api.foo.com/abc"
data = get_data()

return requets.request(
  "POST"
  url=url,
  data=data,
  cookies=self.cookies,
  auth=self.auth,
  timeout=TIMEOUT,
  verify=False
)
```
And the mostly the same for /bar. How could this be done more cleanly without duplicate code?

**Comment by Pierre-Sassoulas** [comment]
> a_lot_of_local_variables = SOME_OBJECT.a_lot_of_local_variables
> with_long_names = SOME_OBJECT.with_long_names
> that_dont_fit = SOME_OBJECT.that_dont_fit
> on_one_line = SOME_OBJECT.on_one_line

In this case you can use ``SOME_OBJECT.x`` directly instead of creating local variables.

> You dont always have control over the function you are calling.

This is a lot more convincing, +1 from me then.

**Comment by ruro** [comment]
> > You dont always have control over the function you are calling.
>
> This is a lot more convincing, +1 from me then.

I am glad that you have seen the light, lol, Buuuut... I'm still gonna try to address your argument just in case.

> In this case you can use SOME_OBJECT.x directly instead of creating local variables.

Once again, IMHO, this is a workaround, not a general fix. You can sometimes use `SOME_OBJECT.x` instead of assigning to a local variable, but this doesn't always work and even in cases where it does work, it isn't always a good idea.

1) Local variables are semantically different from properties on an object. For example, in-place operators and reassignment work differently when the LHS is the name of a local variable vs when it's an attribute access a-la `SOME_OBJECT.x`.

2) Accessing attributes on a global object can be noticeably slower in a tight loop compared to accessing local variables.

3) The name of the attribute on an object is determined by the writer of the class, while the name of a local variable is **local** to the function (duh), allowing you to choose a better, more expressive name (for self-documenting code).

Also, note that this "fix" actually introduces **more** repetition/duplication into your code. Instead of destructuring or `getattr`ing once and assigning to a variable, you are suggesting to repeat the attribute access in every place the variable was originally used. This is almost the opposite of DRY.

In this hypothetical, we started with an incredibly simple operation - a completely normal tuple destructuring (that just **happened** to be spread over multiple lines because of the long variable names and choice of autoformatter/style), but then by trying to "fix" the spurious "duplicate code" warning, we ended up introducing a custom class to store the data instead of just using a regular tuple and then repeating the same attribute access all over the place instead of just using a descriptively named local variable.
